lib/json_schemer/schema.rb

```diff
@@ -0,0 +1,309 @@
+# frozen_string_literal: true
+module JSONSchemer
+  class Schema
+    Context = Struct.new(:instance, :dynamic_scope, :adjacent_results, :short_circuit, :access_mode) do
+      def original_instance(instance_location)
+        Hana::Pointer.parse(Location.resolve(instance_location)).reduce(instance) do |obj, token|
+          obj.fetch(obj.is_a?(Array) ? token.to_i : token)
+        end
+      end
+    end
+
+    include Output
+    include Format::JSONPointer
+
+    DEFAULT_SCHEMA = Draft202012::BASE_URI.to_s.freeze
+    UNKNOWN_KEYWORD_CLASS = Draft202012::Vocab::Core::UnknownKeyword
+    DEFAULT_BASE_URI = URI('json-schemer://schema').freeze
+    DEFAULT_FORMATS = {}.freeze
+    DEFAULT_KEYWORDS = {}.freeze
+    DEFAULT_BEFORE_PROPERTY_VALIDATION = [].freeze
+    DEFAULT_AFTER_PROPERTY_VALIDATION = [].freeze
+    DEFAULT_REF_RESOLVER = proc { |uri| raise UnknownRef, uri.to_s }
+    NET_HTTP_REF_RESOLVER = proc { |uri| JSON.parse(Net::HTTP.get(uri)) }
+    RUBY_REGEXP_RESOLVER = proc { |pattern| Regexp.new(pattern) }
+    ECMA_REGEXP_RESOLVER = proc { |pattern| Regexp.new(EcmaRegexp.ruby_equivalent(pattern)) }
+
+    DEFAULT_PROPERTY_DEFAULT_RESOLVER = proc do |instance, property, results_with_tree_validity|
+      results_with_tree_validity = results_with_tree_validity.select(&:last) unless results_with_tree_validity.size == 1
+      annotations = results_with_tree_validity.to_set { |result, _tree_valid| result.annotation }
+      if annotations.size == 1
+        instance[property] = annotations.first.clone
+        true
+      else
+        false
+      end
+    end
+
+    attr_accessor :base_uri, :meta_schema, :keywords, :keyword_order
+    attr_reader :value, :parent, :root, :parsed
+    attr_reader :vocabulary, :format, :formats, :custom_keywords, :before_property_validation, :after_property_validation, :insert_property_defaults, :property_default_resolver
+
+    def initialize(
```

Comment:
Do all the `options` passed into `JsonSchemer.schema` [here](https://github.com/davishmcclurg/json_schemer/pull/135/files#diff-5a8df3a8be9073d3c94d5d662f070ee071a308147a4c78f6662ff06bc0175ab0R115) all get passed into this `initialize` method? 
I'm wondering if this provides options beyond what's currently in the readme, perhaps an interface that could either currently, or in the future, be used to extend available behavior. 🤔 

For example, I have date properties (type = string, format = date) for which I want configurable minimum and maximum values. As the `minimum` and `maximum` properties are only valid for `number` types, I'm looking at something like the `formatMinimum` (and similar, supported by [AJV](https://ajv.js.org/packages/ajv-formats.html#keywords-to-compare-values-formatmaximum-formatminimum-and-formatexclusivemaximum-formatexclusiveminimum)). 

I'm wondering whether there are ways to extend JSON Schemer 2.0 to add either specific `vocabularies` or custom validations or other such hooks. (the before/after property validation hooks look great, but like they're more designed for data transformations, but perhaps there's a way to use them for other purposes?)

Reply:
> Do all the `options` passed into `JsonSchemer.schema` [here](https://github.com/davishmcclurg/json_schemer/pull/135/files#diff-5a8df3a8be9073d3c94d5d662f070ee071a308147a4c78f6662ff06bc0175ab0R115) all get passed into this `initialize` method?

Yes they do.

> For example, I have date properties (type = string, format = date) for which I want configurable minimum and maximum values. As the `minimum` and `maximum` properties are only valid for `number` types, I'm looking at something like the `formatMinimum` (and similar, supported by [AJV](https://ajv.js.org/packages/ajv-formats.html#keywords-to-compare-values-formatmaximum-formatminimum-and-formatexclusivemaximum-formatexclusiveminimum)).

You could probably do that with the existing `keywords` option, which takes a hash of keywords to callables. It's not documented because I was hoping to come up with a better way to register custom keywords. This PR introduces keyword classes, which might make that easier.

> I'm wondering whether there are ways to extend JSON Schemer 2.0 to add either specific `vocabularies` or custom validations or other such hooks. (the before/after property validation hooks look great, but like they're more designed for data transformations, but perhaps there's a way to use them for other purposes?)

Once this gets merged, I think custom vocabularies will be an interesting option. I haven't taken the time to think it through yet but it seems like it makes sense to go along with custom keywords.